### PR TITLE
Feature/372 wizard testing

### PR DESCRIPTION
**Pull request by RubenFricke**
## Description

Added tests for the wizard.

## Type of change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

1. Validate that all tests are succeeding
2. Test the functionality of the wizard to make sure the CRUD functionality of the data source controller works.

## Link to issue

Closes: #372 
Closes: #385

### CODE REVIEW

File: API/Controllers/WizardController.cs [hunk 1]
```diff
@@ -266,7 +266,7 @@ public class WizardController : ControllerBase
         [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
         public async Task<IActionResult> GetProjectByGuidFromExternalDataSource([FromQuery] string dataSourceGuid,
             [FromQuery] string token,
-            int projectId,
+            string projectId,
```
Comment (JVerbruggen) [inline API/Controllers/WizardController.cs:269]: Why did you change this to a string?

File: API/Extensions/DependencyInjectionExtensions.cs [hunk 2]
```diff
@@ -100,6 +101,8 @@ public static IServiceCollection AddServicesAndRepositories(this IServiceCollect
             services.AddScoped<IWizardPageService, WizardPageService>();
             services.AddScoped<IWizardPageRepository, WizardPageRepository>();
 
+            services.AddScoped<IAssemblyHelper, AssemblyHelper>();
```
Comment (JVerbruggen) [inline API/Extensions/DependencyInjectionExtensions.cs:104]: Why scoped? Is it necessary for each session to get a separate one? If not, why not use a singleton?

File: CHANGELOG.md [hunk 3]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Graduation system for users whose FHICT account becomes unavailable. [#287](https://github.com/DigitalExcellence/dex-backend/issues/287)
 - Added wizard to automatically import project with a public and an oauth flow - [326](https://github.com/DigitalExcellence/dex-backend/issues/326)
+- Added tests for the wizard - [372](https://github.com/DigitalExcellence/dex-backend/issues/372)
```
Comment (JVerbruggen) [inline CHANGELOG.md:17]: Inconsistent '#' before issue number..

File: Models/Defaults/Defaults.cs [hunk 4]
```diff
@@ -135,6 +135,9 @@ public static class Scopes
             [Description("This scope gives write access to the project namespace")]
             public const string AdminProjectWrite = "adminproject:write";
 
+            [Description("This scopes give write access to the wizard page namespace")]
+            public const string WizardPageWrite = "wizardPage:write";
```
Comment (JVerbruggen) [inline Models/Defaults/Defaults.cs:139]: inconsistent naming.. you use camelCase but others do not

File: API/Controllers/WizardController.cs (hunk 1 above)
Comment (RubenFricke) [reply API/Controllers/WizardController.cs:269]: This method will work with external API calls (indirectly). Because some external data sources will get added later, we don't know the specific specifications of that API. Some API's might use integers as id's for their projects, but some other API's might use guids, or even other data types. 

So, therefore I decided to use the string. I can use guids in a string, but also integers. Later in the process, every adaptee can parse this string in the way their API needs to id delivered.

Good question and thanks for the reminder :wink:

File: Services.Tests/03_Services.Tests.csproj [hunk 5]
```diff
@@ -10,13 +10,15 @@
...
     <PackageReference Include="FluentAssertions" Version="5.10.3" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.12" />
     <PackageReference Include="Moq" Version="4.13.1" />
     <PackageReference Include="nunit" Version="3.12.0" />
     <PackageReference Include="NUnit3TestAdapter" Version="3.15.1" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\API\01_API.csproj" />
```
Comment (JVerbruggen) [inline Services.Tests/03_Services.Tests.csproj:21]: Why does the services project need a project reference to the API layer??

File: Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+            loader = new DataProviderLoader(factoryMock.Object, repositoryMock.Object, assemblyHelperMock.Object ,mapper);
+        }
+
+
+        [Test]
+        public async Task GetAllDataSources_GoodFlow()
+        {
+            // Arrange
+            returnedTypesFromAssembly = new[]
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs:70]: Unclear what this is for

Comment (JVerbruggen) [reply Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs:70]: Why is it a global var?

File: Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+            MapperConfiguration mappingConfig = new MapperConfiguration(mc => mc.AddProfile(new MappingProfile()));
+            mapper = mappingConfig.CreateMapper();
+
+            loader = new DataProviderLoader(factoryMock.Object, repositoryMock.Object, assemblyHelperMock.Object ,mapper);
+        }
+
+
+        [Test]
+        public async Task GetAllDataSources_GoodFlow()
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs:67]: Can you maybe add a description how the test accomplishes the good flow?

Comment (JVerbruggen) [reply Services.Tests/ExternalDataProviders/DataProviderLoaderTest.cs:67]: Also maybe do this for the other tests? Would be nice for rookies like me to understand it! :)

File: Services.Tests/ExternalDataProviders/DataProviderServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+        [Test]
+        public async Task GetAllProjectsInAuthFlow_GoodFlow([ProjectDataSource(50)] IEnumerable<Project> projects)
+        {
+            // Arrange
+            dataSourceMock.As<IAuthorizedDataSourceAdaptee>().Setup(_ => _.GetAllProjects(It.IsAny<string>()))
+                          .ReturnsAsync(projects);
+
+            // Act
+            Action act = () => service.GetAllProjects(It.IsAny<string>(), It.IsAny<string>(), true);
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataProviderServiceTest.cs:80]: what is going on here? (not breakfast)

File: Services.Tests/ExternalDataProviders/DataSources/JsFiddle/JsFiddleDataSourceResourceResultDataSourceAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        public IEnumerable GetData(IParameterInfo parameter)
+        {
+            if(amountToGenerate <= 1)
+            {
+                return new[] {fakeDataGenerator.Generate()};
+            }
+            List<JsFiddleDataSourceResourceResult> projects = fakeDataGenerator.GenerateRange(amountToGenerate)
+                                                                             .ToList();
+            return new [] {projects};
+        }
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataSources/JsFiddle/JsFiddleDataSourceResourceResultDataSourceAttribute.cs:72]: You're doing a lot of the same thing here.. Maybe there's a way to create an abstract class above these? (See other identical comments I left which indicate what classes I'm talking about)

File: Services.Tests/ExternalDataProviders/DataSources/Gitlab/GitlabDataSourceUserResourceResultDataSourceAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        public IEnumerable GetData(IParameterInfo parameter)
+        {
+            if(amountToGenerate <= 1)
+            {
+                return new[] {fakeDataGenerator.Generate()};
+            }
+            List<GitlabDataSourceUserResourceResult> projects = fakeDataGenerator.GenerateRange(amountToGenerate)
+                                                                             .ToList();
+            return new [] {projects};
+        }
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataSources/Gitlab/GitlabDataSourceUserResourceResultDataSourceAttribute.cs:72]: You're doing a lot of the same thing here.. Maybe there's a way to create an abstract class above these? (See other identical comments I left which indicate what classes I'm talking about)

File: Services.Tests/ExternalDataProviders/DataSources/Github/GithubDataSourceResourceResultDataSourceAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        public IEnumerable GetData(IParameterInfo parameter)
+        {
+            if(amountToGenerate <= 1)
+            {
+                return new[] {fakeDataGenerator.Generate()};
+            }
+            List<GithubDataSourceResourceResult> projects = fakeDataGenerator.GenerateRange(amountToGenerate)
+                                                                             .ToList();
+            return new [] {projects};
+        }
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataSources/Github/GithubDataSourceResourceResultDataSourceAttribute.cs:70]: You're doing a lot of the same thing here.. Maybe there's a way to create an abstract class above these? (See other identical comments I left which indicate what classes I'm talking about)

File: Services.Tests/ExternalDataProviders/DataSources/Gitlab/GitlabDataSourceResourceResultDataSourceAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        /// <returns>The generated data.</returns>
+        public IEnumerable GetData(IParameterInfo parameter)
+        {
+            if(amountToGenerate <= 1)
+            {
+                return new[] {fakeDataGenerator.Generate()};
+            }
+            List<GitlabDataSourceResourceResult> projects = fakeDataGenerator.GenerateRange(amountToGenerate)
+                                                                             .ToList();
+            return new [] {projects};
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/DataSources/Gitlab/GitlabDataSourceResourceResultDataSourceAttribute.cs:70]: You're doing a lot of the same thing here.. Maybe there's a way to create an abstract class above these? (See other identical comments I left which indicate what classes I'm talking about)

File: Services.Tests/ExternalDataProviders/GithubDataSourceAdapteeTest.cs [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+                                                          {"App:DataSources:Github", "valid_access_token"}
+                                                      };
+
+            ConfigurationMock = new ConfigurationBuilder()
+                                .AddInMemoryCollection(inMemoryConf)
+                                .Build();
+        }
+
+        [Test]
+        public async Task FetchAllGithubProjects_GoodFlow(
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/GithubDataSourceAdapteeTest.cs:57]: Not clear to rookies what is happening...

Comment (JVerbruggen) [reply Services.Tests/ExternalDataProviders/GithubDataSourceAdapteeTest.cs:57]: (see other comment)

File: Services.Tests/ExternalDataProviders/JsFiddleDataSourceAdapteeTest.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace Services.Tests.ExternalDataProviders
+{
+
+    [TestFixture]
+    public class JsFiddleDataSourceAdapteeTest : AdapteeTest<IJsFiddleDataSourceAdaptee>
+    {
+
+        [Test]
+        public async Task FetchAllFiddlesFromUser_GoodFlow(
```
Comment (JVerbruggen) [inline Services.Tests/ExternalDataProviders/JsFiddleDataSourceAdapteeTest.cs:40]: Not clear what is happening (see other comment)

File: Services.Tests/WizardPageServiceTest.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    public class WizardPageServiceTest : ServiceTest<WizardPage, WizardPageService, IWizardPageRepository>
+    {
+
+        /// <summary>
+        ///     This methods tests the ValidateWizardPagesExists method in a good flow where the wizard pages do exist.
+        /// </summary>
+        /// <param name="wizardPages">A collection of 10 generated wizard pages.</param>
+        /// <returns></returns>
+        [Test]
+        public async Task ValidateWizardPagesExist_WizardPageExists([WizardPageDataSource(10)] IEnumerable<WizardPage> wizardPages)
```
Comment (JVerbruggen) [inline Services.Tests/WizardPageServiceTest.cs:48]: Not clear what is happening (see other comment)

File: Services/ExternalDataProviders/Helpers/AssemblyHelper.cs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    public interface IAssemblyHelper
+    {
+
+        Type[] RetrieveTypesFromExecutingAssemblyFolderFiles();
+
+        Type[] RetrieveTypesFromExecutingAssemblyFolderFolderByInterface(Type @interface);
+
+    }
+
+    public class AssemblyHelper : IAssemblyHelper
```
Comment (JVerbruggen) [inline Services/ExternalDataProviders/Helpers/AssemblyHelper.cs:35]: What is this class used for?

File: Services/ExternalDataProviders/JsFiddleDataSourceAdaptee.cs [hunk 17]
```diff
@@ -31,7 +31,18 @@
 namespace Services.ExternalDataProviders
 {
 
-    public class JsFiddleDataSourceAdaptee : IPublicDataSourceAdaptee
+    public interface IJsFiddleDataSourceAdaptee : IPublicDataSourceAdaptee
+    {
+
+        Task<IEnumerable<JsFiddleDataSourceResourceResult>> FetchAllFiddlesFromUser(string username);
```
Comment (JVerbruggen) [inline Services/ExternalDataProviders/JsFiddleDataSourceAdaptee.cs:37]: Missing descriptions here

**Review by JVerbruggen — CHANGES_REQUESTED** [review]
I left some suggestions and questions in my review.. Overall it looks really well done, just some finishing touches probably! I did not test if unittests & postman tests validate, but I'm sure you tested that.

**Review by DamienKusters — APPROVED** [review]

File: API/Extensions/DependencyInjectionExtensions.cs (hunk 2 above)
Comment (RubenFricke) [reply API/Extensions/DependencyInjectionExtensions.cs]: You're right. I should have chosen singleton. The only reason I initially choose to go for scoped is to support dependency injection for our repositories, but after a closer look, I don't think this is necessary.

**Review by DamienKusters — APPROVED** [review]

**Review by DamienKusters — DISMISSED** [review]
No comments

File: Services.Tests/ExternalDataProviders/GithubDataSourceAdapteeTest.cs [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+
+namespace Services.Tests.ExternalDataProviders
+{
+
+    /// <summary>
+    ///     GithubDataSourceAdapteeTest
+    /// </summary>
+    /// <seealso cref="IGithubDataSourceAdaptee" />
+    [TestFixture]
+    public class GithubDataSourceAdapteeTest : AdapteeTest<IGithubDataSourceAdaptee>
```
Comment (niraymak) [inline Services.Tests/ExternalDataProviders/GithubDataSourceAdapteeTest.cs:41]: I don't know if you did this on purpose. I used Live Unit Testing to see the coverage and metioned that the following methods in GithubDataSourceAdaptee.cs are not covered:

GetTokens
GetAllProjects
GetProjectById
GetAllPublicProjects
GetPublicProjectFromUri
GetPublicProjectById

File: Services.Tests/ExternalDataProviders/GitlabDataSourceAdapteeTest.cs [hunk 19]
```diff
@@ -0,0 +1,417 @@
...
+
+namespace Services.Tests.ExternalDataProviders
+{
+
+    /// <summary>
+    ///     GitlabDataSourceAdapteeTest
+    /// </summary>
+    /// <seealso cref="IGitlabDataSourceAdaptee" />
+    [TestFixture]
+    public class GitlabDataSourceAdapteeTest : AdapteeTest<IGitlabDataSourceAdaptee>
```
Comment (niraymak) [inline Services.Tests/ExternalDataProviders/GitlabDataSourceAdapteeTest.cs:40]: The following methods in GitlabDataSourceAdaptee.cs are not covered:

GetTokens
GetAllProjects
GetProjectById
GetAllPublicProjects
GetPublicProjectFromUri
GetPublicProjectById

File: Services.Tests/ExternalDataProviders/JsFiddleDataSourceAdapteeTest.cs [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+namespace Services.Tests.ExternalDataProviders
+{
+
+    /// <summary>
+    ///     JsFiddleDataSourceAdapteeTest
+    /// </summary>
+    /// <seealso cref="IJsFiddleDataSourceAdaptee" />
+    [TestFixture]
+    public class JsFiddleDataSourceAdapteeTest : AdapteeTest<IJsFiddleDataSourceAdaptee>
```
Comment (niraymak) [inline Services.Tests/ExternalDataProviders/JsFiddleDataSourceAdapteeTest.cs:40]: Not covered in JsFiddleDataSourceAdaptee.cs

GetAllPublicProjects
GetPublicProjectFromUri
GetPublicProjectById

**Review by niraymak — CHANGES_REQUESTED** [review]
Nice tests! I mentioned a few methods that weren't covered and we already discussed that some weren't necessary to cover. If the ones that need to be covered, I will approve :)

**Review by niraymak — APPROVED** [review]
Everything seems to be covered! 
Tested CRUD functionality of DataSources and that seems to work aswell :)
Good job! 🥇

**Comment by niraymak** [comment]
Just tested this in combination with frontend implementation and found out that GetProject endpoint was quite slow sometimes.
Also tested with develop and did not notice it there. 
image

File: API/Extensions/DependencyInjectionExtensions.cs [hunk 21]
```diff
@@ -102,30 +103,31 @@ public static IServiceCollection AddServicesAndRepositories(this IServiceCollect
...
             services.AddScoped<IWizardPageRepository, WizardPageRepository>();
 
+            services.AddScoped<IUserProjectLikeService, UserProjectLikeService>();
+            services.AddScoped<IUserProjectLikeRepository, UserProjectLikeRepository>();
+
+            services.AddScoped<ICallToActionOptionService, CallToActionOptionService>();
+            services.AddScoped<ICallToActionOptionRepository, CallToActionOptionRepository>();
+
+            services.AddSingleton<Queries>();
+            services.AddScoped<ITaskPublisher, TaskPublisher>();
```
Comment (niraymak) [inline API/Extensions/DependencyInjectionExtensions.cs:113]: This should be a AddSingleton

**Review by niraymak — CHANGES_REQUESTED** [review]

**Review by niraymak — APPROVED** [review]

**Review by JVerbruggen — APPROVED** [review]
Lekker bezig Ruben. Van mij krijg je een approve.
